File: gomoku.py
import json, unittest
from flask import Flask, render_template, request, session
from flask_socketio import SocketIO, emit, join_room, leave_room, close_room, rooms, disconnect

app=Flask(__name__)
app.config['DEBUG']=True
app.config['SECRET_KEY']='secret'
socketio=SocketIO(app)

class chess(object):
	variable=0
	username=[]
	loser=''
	chessBar=[[0 for col in range(15)]for row in range(15)]

def clear():
	chess.variable=0
	chess.username=[]
	chess.loser=''
	chess.chessBar=[[0 for col in range(15)]for row in range(15)]


def clock(key):
	if key==0:
		return(1)
	if key==1:
		return(0)
	else:
		raise ValueError

def judgeWin(varification,coordinate):
	judge=[]
	for i in range(4):
		judge.append(fiveConnection(i,varification,coordinate))
	print(judge)
	for x in judge:
		if x:
			return(1)
	return(0)

def fiveConnection(type,varification,coordinate):
	count=0
	if type==0:
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (x<15 and chess.chessBar[x][y]==varification):
				count+=1
				x+=1
			else:
				break
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (x>=0 and chess.chessBar[x][y]==varification):
				count+=1
				x-=1
			else:
				break
	elif type==1:
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (x<15 and y<15 and chess.chessBar[x][y]==varification):
				count+=1
				x+=1
				y+=1
			else:
				break
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (x>=0 and y>=0 and chess.chessBar[x][y]==varification):
				count+=1
				x-=1
				y-=1
			else:
				break
	elif type==2:
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (y<15 and chess.chessBar[x][y]==varification):
				count+=1
				y+=1
			else:
				break
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (y>=0 and chess.chessBar[x][y]==varification):
				count+=1
				y-=1
			else:
				break
	elif type==3:
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (x>=0 and y<15 and chess.chessBar[x][y]==varification):
				count+=1
				x-=1
				y+=1
			else:
				break
		x=coordinate[0]
		y=coordinate[1]
		while True:
			if (x<15 and y>=0 and chess.chessBar[x][y]==varification):
				count+=1
				x+=1
				y-=1
			else:
				break
	print(count)
	if count>5:
		return(1)
	else:
		return(0)


@app.route('/')
def index():
	return render_template('index.html')


@app.route('/gomoku')
def gomoku():
	return render_template('gomoku.html')


@socketio.on('connect',namespace='/')
def handler_connect():
	emit('response', {'data': 'connection established!'})

@socketio.on('prepare')
def handler_prepare(message):
	if (len(chess.username)==0):
		if message['data']:
			chess.username.append(message['data'])
			emit('response', {'data': chess.username[0]+' is prepared.'},broadcast=True)
	elif (len(chess.username)==1):
		if message['data'] and message['data']!=chess.username[0]:
			chess.username.append(message['data'])
			emit('response', {'data': chess.username[1]+' is prepared.'},broadcast=True)
			msgStart=json.dumps({'start':1,'player1':chess.username[0],'player2':chess.username[1]})
			msgRight=json.dumps({'right':chess.username[0],'image':'../static/gomoku/black.png'})
			emit('start', {'data': msgStart},broadcast=True)
			emit('right', {'data': msgRight},broadcast=True)
	else:
		if (len(chess.username)>1 and len(chess.username)<10):
			if message['data']:
				chess.username.append(message['data'])
			emit('response', {'data': 'this room is full.'})

@socketio.on('chess')
def handler_chess(message):#{'user':'jack','image':'../static/gomoku/black.png','x':6,'y':5}
	obj=json.loads(message['data'])
	if obj['user']==chess.username[chess.variable] and chess.chessBar[obj['x']][obj['y']]==0:
		emit('coordinate', {'data': message['data']},broadcast=True)
		chess.chessBar[obj['x']][obj['y']]=chess.variable+1
		dot=[obj['x'],obj['y']]
		if judgeWin(chess.variable+1,dot):
			msgResult=json.dumps({'result': chess.username[chess.variable]})
			print(msgResult)
			emit('result',{'data': msgResult},broadcast=True)
		else:
			chess.variable=clock(chess.variable)
			if chess.variable:
				image='../static/gomoku/white.png'
			else:
				image='../static/gomoku/black.png'
			msgRight=json.dumps({'right': chess.username[chess.variable],'image':image})
			emit('right', {'data': msgRight},broadcast=True)

@socketio.on('message')
def handler_message(message):
	try:
		obj=json.loads(message['data'])#{"user":"jackson","msg":"hello,everyone"}
		print(obj['user']+':'+obj['msg'])
	except ValueError as e:
		print('It is not a chat.')
	finally:
		emit('response',{'data':message['data']},broadcast=True)

@socketio.on('clear')
def handler_clear(message):
	obj=json.loads(message['data'])
	if(obj['user'] in chess.username):
		emit('clear',broadcast=True)
		chess.clear()

@socketio.on('restart')
def handler_restart(message):
	if(message['data']==chess.username[0]):
		chess.loser=chess.username[0]
	elif(message['data']==chess.username[1]):
		chess.loser=chess.username[1]
	print('chess.loser: '+chess.loser)
	emit('restartConfirm',{'data':chess.loser},broadcast=True)

@socketio.on('restartConfirm')
def handler_restartConfirm(message):
	obj=json.loads(message['data'])
	if(obj['user'] in chess.username and obj['user']!=chess.loser):
		emit('restartConfirm',{'data':message['data']},broadcast=True)
		if (obj['confirm']==1):
			print('game will be clear.')
			chess.clear()

@socketio.on('getUser')
def handler_getUser():
	obj=json.dumps(chess.username)
	emit('response',{'data':obj})

@socketio.on('getVariable')
def handler_getVariable():
	emit('response',{'data':chess.variable})


if __name__=='__main__':
    socketio.run(app,host='0.0.0.0',port=5000)

File: app/static/gomoku/gomoku.js
//游戏逻辑
//创建一个五子棋对象，并初始化
var chess=new Object();
//初始化本地用户
chess.user=prompt('What is your name?');
$('#player1').html('<p>玩家：'+chess.user+'</p>');
function initial(){
	//棋子的序号，描述落子的先后顺序
	chess.code=1;
	//你的对手
	chess.opponent=''
	//对应image监听器
	chess.image='';
	//对应start监听器
	chess.start=0;
	//对应right监听器
	chess.right='';
	//对应coordinate监听器
	chess.coordinate=[];
	//对应result监听器
	chess.result='';
	//对应restart监听器
	chess.loser='';
	//清除场上的棋子
	list=$('span');
	log('Now,it is a new game.');
	for(var i=0;i<list.length;i++){
		list[i].parentNode.style.backgroundImage='';
		list[i].parentNode.innerHTML='';
	}
}
initial();
//获取元素坐标的函数
function getCoordinate(obj){
	var chessBar=document.querySelectorAll('.chess-line');
	var parentElement=obj.parentElement;
	var i=0,j=0;
	for (i = 0; obj; i++) {
		obj=obj.previousElementSibling;
	}
	for (j = 0; parentElement; j++) {
		parentElement=parentElement.previousElementSibling;
	}
	return [i-1,j-1];
}
//通过坐标获取元素的函数
function getCell(coordinate){
	var x=-1,y=-1;
	x=coordinate[0];
	y=coordinate[1];
	var chessBar=document.querySelectorAll('.chess-line');
	var obj=chessBar[y].childNodes[2*x+1];
	return obj;
}
//log生成函数
function log(msg) {
	document.getElementById('log').innerHTML='<p>'+msg+'</p>'+document.getElementById('log').innerHTML;
}
//准备消息
function prepare(){
	socket.emit('prepare',{data: chess.user});//jack
}
//发送落子消息
function play(obj){
	if(chess.start&&chess.right===chess.user){
		var coordinate=getCoordinate(obj);
		var chessJson=new Object();
		chessJson.user=chess.user;
		chessJson.image=chess.image;
		chessJson.x=coordinate[0];
		chessJson.y=coordinate[1];
		var chessData=JSON.stringify(chessJson);
		socket.emit('chess',{data: chessData});//{'user':'jack','image':'../static/gomoku/black.png','x':6,'y':5}
	}
}
//聊天模块
//注册键盘事件
document.onkeydown = function(e) {
	//捕捉回车事件
	var ev = (typeof event!= 'undefined') ? window.event : e;
	if(ev.keyCode == 13) {
		send();
	}
}
function send() {
	if(!document.getElementById('text').value){
		alert('please enter something.');
		return false;
	}
	var msg='';
	msg=document.getElementById('text').value;
	var message=new Object();
	message.user=chess.user;
	message.msg=msg;
	var msgSend=JSON.stringify(message);
	socket.emit('message', {data: msgSend});
	document.getElementById('text').value='';
};
//请求重来
function restart(){
	if (chess.start===1) {
		var varification=confirm('你希望认输并开始下一局吗？如果你的对手不同意则不会生效。')
		if(varification){
			chess.loser=chess.user;
			var message=new Object();
			message.loser=chess.loser;
			var msgRestart=JSON.stringify(message);
			socket.emit('restart',{data: msgRestart});//{'loser':'jack'}
			console.log(chess.user);
		}
	}
	if(chess.start===0){
		var message=new Object();
		message.user=chess.user;
		var msgClear=JSON.stringify(message);
		socket.emit('clear', {data: msgClear});
	}
}

function gameHelp(){
	$("#gameHelp").slideToggle();
}
function getUser(){
	socket.emit('getUser');
}
function getVariable(){
	socket.emit('getVariable');
}